File: apps/api/app/models/activity.py
"""Activity and presence models — activity feed and user presence.

FM-058: Activity feed entries (audit-like stream per project/workspace).
FM-059: Real-time user presence tracking.
"""

import enum
import uuid
from datetime import datetime

from sqlalchemy import String, Text, DateTime, Enum, ForeignKey, func
from sqlalchemy.dialects.postgresql import UUID, JSON
from sqlalchemy.orm import Mapped, mapped_column

from app.db.base_class import Base


class ActivityType(str, enum.Enum):
    PROJECT_CREATED = "project_created"
    PROJECT_UPDATED = "project_updated"
    RUN_STARTED = "run_started"
    RUN_COMPLETED = "run_completed"
    RUN_FAILED = "run_failed"
    TASK_COMPLETED = "task_completed"
    ARTIFACT_CREATED = "artifact_created"
    MEMBER_ADDED = "member_added"
    MEMBER_REMOVED = "member_removed"
    APPROVAL_REQUESTED = "approval_requested"
    APPROVAL_DECIDED = "approval_decided"
    ESCALATION_TRIGGERED = "escalation_triggered"
    PATCH_PROPOSED = "patch_proposed"
    PR_CREATED = "pr_created"
    COMMENT = "comment"


class ActivityFeedEntry(Base):
    __tablename__ = "activity_feed_entries"

    id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), primary_key=True, default=uuid.uuid4
    )
    actor_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("users.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    activity_type: Mapped[ActivityType] = mapped_column(
        Enum(ActivityType, name="activity_type"),
        nullable=False,
        index=True,
    )
    summary: Mapped[str] = mapped_column(String(500), nullable=False)

    project_id: Mapped[uuid.UUID | None] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("projects.id", ondelete="SET NULL"),
        nullable=True,
        index=True,
    )
    workspace_id: Mapped[uuid.UUID | None] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("workspaces.id", ondelete="SET NULL"),
        nullable=True,
    )
    resource_type: Mapped[str | None] = mapped_column(String(50), nullable=True)
    resource_id: Mapped[uuid.UUID | None] = mapped_column(
        UUID(as_uuid=True), nullable=True
    )

    metadata_: Mapped[dict | None] = mapped_column(
        "metadata", JSON, nullable=True
    )

    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now(), nullable=False
    )

    def __repr__(self) -> str:
        return f"<Activity {self.activity_type.value} by={self.actor_id}>"


class UserPresence(Base):
    __tablename__ = "user_presences"

    id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), primary_key=True, default=uuid.uuid4
    )
    user_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("users.id", ondelete="CASCADE"),
        nullable=False,
        unique=True,
        index=True,
    )
    status: Mapped[str] = mapped_column(
        String(20), default="online", nullable=False
    )
    current_resource_type: Mapped[str | None] = mapped_column(
        String(50), nullable=True
    )
    current_resource_id: Mapped[uuid.UUID | None] = mapped_column(
        UUID(as_uuid=True), nullable=True
    )

    last_seen_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now(), nullable=False
    )

    def __repr__(self) -> str:
        return f"<Presence user={self.user_id} status={self.status}>"

File: packages/schemas/src/activity.ts
/** Activity and presence types matching the backend schemas. */

export type ActivityType =
  | "project_created"
  | "project_updated"
  | "run_started"
  | "run_completed"
  | "run_failed"
  | "task_completed"
  | "artifact_created"
  | "member_added"
  | "member_removed"
  | "approval_requested"
  | "approval_decided"
  | "escalation_triggered"
  | "patch_proposed"
  | "pr_created"
  | "comment";

export interface ActivityFeedEntry {
  id: string;
  actor_id: string;
  activity_type: ActivityType;
  summary: string;
  project_id: string | null;
  workspace_id: string | null;
  resource_type: string | null;
  resource_id: string | null;
  metadata_: Record<string, unknown> | null;
  created_at: string;
}

export interface ActivityFeedList {
  items: ActivityFeedEntry[];
  total: number;
}

export interface Presence {
  id: string;
  user_id: string;
  status: string;
  current_resource_type: string | null;
  current_resource_id: string | null;
  last_seen_at: string;
}

export interface PresenceList {
  items: Presence[];
  total: number;
}

export interface UserContext {
  user_id: string;
  status: string;
  last_seen_at: string | null;
  current_resource_type: string | null;
  workspace_memberships: number;
  project_memberships: number;
}
